### PR TITLE
Improve the help text

### PR DESCRIPTION
**Pull request by jgoshi**
@piotrpMSFT @livarcocc @krwq @blackdwarf 

Fixes issue #4968 

Also removing the --force option from dotnet add and refactoring to use sub-commands since we already have that infrastructure in place. This will help consistency across all our commands.

### CODE REVIEW

**Comment by jgoshi** [comment]
test RHEL7.2 x64 Release Build

**Comment by jgoshi** [comment]
test Fedora23 x64 Debug Build

File: src/Microsoft.DotNet.Cli.Utils/Constants.cs [hunk 1]
```diff
@@ -48,5 +48,7 @@ public static class Constants
         };
 
         public static readonly string MSBUILD_EXE_PATH = "MSBUILD_EXE_PATH";
+
+        public static readonly string ProjectOrSolutionArgumentName = "<PROJECT_OR_SOLUTION>";
```
Comment (krwq) [inline src/Microsoft.DotNet.Cli.Utils/Constants.cs]: I'm not sure the underscores are needed considering you wrap it with `<>` cc: @blackdwarf

File: src/dotnet/CommandLine/CommandLineApplication.cs [hunk 2]
```diff
@@ -233,7 +239,7 @@ public int Execute(params string[] args)
                     option = null;
                 }
 
-                if (!processed && arguments == null)
+                if (!processed)
```
Comment (krwq) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: merge this if with the next one

File: src/dotnet/CommandLine/CommandLineApplication.cs [hunk 3]
```diff
@@ -334,7 +340,15 @@ public void ShowHelp(string commandName = null)
             for (var cmd = this; cmd != null; cmd = cmd.Parent)
             {
                 cmd.IsShowingInformation = true;
-                headerBuilder.Insert(6, string.Format(" {0}", cmd.Name));
+                if (cmd != this && cmd.Arguments.Any())
+                {
+                    var args = string.Join(" ", cmd.Arguments.Select(arg => arg.Name));
+                    headerBuilder.Insert(6, string.Format(" {0} {1}", cmd.Name, args));
```
Comment (krwq) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: should this join all the parents' names too before cmd.Name? I'm not sure what the cmd.Parent is in this context but if it's what the name suggest it seems to me that we will get repetition and invalid command invokation in some cases

File: src/dotnet/CommandLine/CommandLineApplication.cs [hunk 4]
```diff
@@ -364,18 +378,36 @@ public void ShowHelp(string commandName = null)
...
 
-                argumentsBuilder.AppendLine();
-                argumentsBuilder.AppendLine("Arguments:");
-                var maxArgLen = MaxArgumentLength(target.Arguments);
-                var outputFormat = string.Format("  {{0, -{0}}}{{1}}", maxArgLen + 2);
-                foreach (var arg in target.Arguments)
+                    if (argumentsBuilder.Length == 0)
+                    {
+                        argumentsBuilder.AppendLine();
+                        argumentsBuilder.AppendLine("Arguments:");
```
Comment (krwq) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: Localize

File: src/dotnet/commands/dotnet-add/Program.cs [hunk 5]
```diff
@@ -14,34 +14,45 @@
...
+            AddProjectToProjectReferenceCommand.CreateApplication,
         };
 
         public static int Run(string[] args)
         {
-            var cmd = new AddCommand();
-            return cmd.Start(args);
+            DebugHelper.HandleDebugSwitch(ref args);
+
+            CommandLineApplication app = new CommandLineApplication(throwOnUnexpectedArg: false)
```
Comment (krwq) [inline src/dotnet/commands/dotnet-add/Program.cs]: is parent command no longer passing anything? is the syntax still the same?

File: src/Microsoft.DotNet.Cli.Utils/Constants.cs (hunk 1 above)
Comment (jgoshi) [inline src/Microsoft.DotNet.Cli.Utils/Constants.cs]: This is what shows up in the usage help text. I believe that's how we write things out right?

File: src/Microsoft.DotNet.Cli.Utils/PathUtility.cs [hunk 6]
```diff
@@ -257,5 +257,16 @@ public static string GetFullPath(string path)
 
             return Path.GetFullPath(path);
         }
+
+        public static string GetCurrentDirectoryWithDirSeparator()
```
Comment (jgoshi) [inline src/Microsoft.DotNet.Cli.Utils/PathUtility.cs]: I prefer the "WithDirSeparator" because it's clear without looking at the implementation what the method is doing.

File: src/dotnet/CommandLine/CommandLineApplication.cs (hunk 3 above)
Comment (jgoshi) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: This is the convention for how things appear to be named. In this specific case, parent name is "dotnet add", the parent argument is &lt;PROJECT_OR_SOLUTION&gt; and the subcommand name is for example "p2p"

**Comment by krwq** [comment]
I went ahead and tested your branch. Here are the issues I found:
- `dotnet add dsfsf sdf sdfdsdf p2p` does not display anything - not even an error
- `dotnet add p2p asdasd` - displays an error encouraging to use `--force` switch
- `dotnet add p2p` - tells to use `dotnet add --help` for more help and displays Usage at the same time

I'm not sure if it makes sense to fix any potential issues which could happen when there is another level of subcommands so other than above lgtm.

File: src/dotnet/CommandLine/CommandLineApplication.cs [hunk 7]
```diff
@@ -364,18 +378,36 @@ public void ShowHelp(string commandName = null)
...
+                }
+            }
+
+            for (var cmd = target; cmd != null; cmd = cmd.Parent)
+            {
+                if (cmd.Arguments.Any())
                 {
-                    argumentsBuilder.AppendFormat(outputFormat, arg.Name, arg.Description);
-                    argumentsBuilder.AppendLine();
+                    var outputFormat = string.Format("  {{0, -{0}}}{{1}}", maxArgLen + 2);
```
Comment (jgoshi) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: Will look into fixing this.

Comment (jgoshi) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: Fixed this.

**Comment by jgoshi** [comment]
@MattGertz This is work to improve help text with the dotnet add,list,remove commands.

**Comment by MattGertz** [comment]
@piotrpMSFT @livarcocc Can one of you review as well (policy is two reviews total), and then tag me.

**Comment by jgoshi** [comment]
test CentOS7.1 x64 Debug Build

**Comment by jgoshi** [comment]
test Windows_NT x64 Release Build

**Comment by jgoshi** [comment]
test CentOS7.1 x64 Debug Build

File: src/Microsoft.DotNet.Cli.Utils/PathUtility.cs (hunk 6 above)
Comment (livarcocc) [inline src/Microsoft.DotNet.Cli.Utils/PathUtility.cs]: We already have this: https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.Cli.Utils/PathUtility.cs#L33.

File: src/dotnet/CommandLine/CommandLineApplication.cs (hunk 3 above)
Comment (livarcocc) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: I know this is not something you changed, but this magic number "6" seems pretty bad. I have no idea what it means.

Comment (krwq) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: I don't think it matters that much - this is formatting code and most likely you don't need to know what this means, I'd rather not see the magic variable in a constant files in this case...

Comment (jgoshi) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: It comes from this: var headerBuilder = new StringBuilder("Usage:"); I can change the magic number 6 to be headerBuilder.Length

Comment (jgoshi) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: sorry, I mean cache the length of Usage: and use that

File: src/dotnet/CommandLine/CommandLineApplication.cs [hunk 8]
```diff
@@ -334,7 +340,15 @@ public void ShowHelp(string commandName = null)
             for (var cmd = this; cmd != null; cmd = cmd.Parent)
             {
                 cmd.IsShowingInformation = true;
-                headerBuilder.Insert(6, string.Format(" {0}", cmd.Name));
+                if (cmd != this && cmd.Arguments.Any())
```
Comment (livarcocc) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: Any chance we can break the pieces of this method into smaller, aptly named methods and invoke them here? It is somewhat hard to follow what exactly is going on in this method because there is very little abstraction.

File: src/dotnet/CommandLine/CommandLineApplication.cs (hunk 3 above)
Comment (krwq) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: I think you will need to do it anyways when you add localization to this class so make sure it happens as part of this or follow-up PR

File: src/dotnet/commands/dotnet-remove/Program.cs [hunk 9]
```diff
@@ -14,34 +14,50 @@
...
         {
-            ["p2p"] = RemoveProjectToProjectReferenceCommand.Run,
+            RemoveProjectToProjectReferenceCommand.CreateApplication,
         };
 
         public static int Run(string[] args)
         {
-            var cmd = new RemoveCommand();
-            return cmd.Start(args);
+            DebugHelper.HandleDebugSwitch(ref args);
```
Comment (livarcocc) [inline src/dotnet/commands/dotnet-remove/Program.cs]: This is the third time this pattern is repeated. Should we extract it and vary only what needs to be varied? Seems like the name, full name and builtincommands?

File: test/dotnet-add-p2p.Tests/GivenDotnetAddP2P.cs [hunk 10]
```diff
@@ -570,44 +580,6 @@ public void WhenPassedMultipleRefsAndOneOfthemDoesNotExistItCancelsWholeOperatio
...
-            cmd.Should().Pass();
-            cmd.StdOut.Should().Contain("added to the project");
-            cmd.StdErr.Should().BeEmpty();
-            var csproj = lib.CsProj();
-            csproj.NumberOfItemGroupsWithoutCondition().Should().Be(noCondBefore + 1);
-            csproj.NumberOfProjectReferencesWithIncludeContaining(nonExisting).Should().Be(1);
-        }
-
-        [Fact]
-        public void WhenPassedReferenceIsUsingSlashesItNormalizesItToBackslashes()
```
Comment (livarcocc) [inline test/dotnet-add-p2p.Tests/GivenDotnetAddP2P.cs]: Is this still a scenario? Albeit without --force?

File: src/dotnet/commands/dotnet-remove/Program.cs (hunk 9 above)
Comment (krwq) [inline src/dotnet/commands/dotnet-remove/Program.cs]: Sounds like DispatchCommand should come back in the end under different form :)

File: test/dotnet-add-p2p.Tests/GivenDotnetAddP2P.cs (hunk 10 above)
Comment (krwq) [inline test/dotnet-add-p2p.Tests/GivenDotnetAddP2P.cs]: It is still a valid scenario

File: src/Microsoft.DotNet.Cli.Utils/Constants.cs (hunk 1 above)
Comment (piotrpMSFT) [inline src/Microsoft.DotNet.Cli.Utils/Constants.cs]: @blackdwarf ?

File: src/dotnet/CommandLine/CommandLineApplication.cs (hunk 4 above)
Comment (piotrpMSFT) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: Thanks for opening the issue

File: src/dotnet/CommandLine/CommandLineApplication.cs (hunk 8 above)
Comment (jgoshi) [inline src/dotnet/CommandLine/CommandLineApplication.cs]: Sure will do.

File: src/dotnet/commands/dotnet-remove/Program.cs (hunk 9 above)
Comment (jgoshi) [inline src/dotnet/commands/dotnet-remove/Program.cs]: Sure :)

File: test/dotnet-add-p2p.Tests/GivenDotnetAddP2P.cs (hunk 10 above)
Comment (jgoshi) [inline test/dotnet-add-p2p.Tests/GivenDotnetAddP2P.cs]: Ok will add it back without the --force

**Comment by jgoshi** [comment]
test CentOS7.1 x64 Debug Build

**Comment by jgoshi** [comment]
test OSX x64 Release Build

**Comment by jgoshi** [comment]
test Windows_NT x86 Debug Build

**Comment by jgoshi** [comment]
test OSX x64 Release Build

**Comment by jgoshi** [comment]
@piotrpMSFT @livarcocc are you guys ok with this version? For the &lt;PROJECT_OR_SOLUTION&gt; comment I'm not sure what this is about. Basically this is the string that appears in the help text usage. And @blackdwarf had suggested that I use that instead of &lt;object&gt;

**Comment by blackdwarf** [comment]
@jgoshi yep, I suggested that because it is clearer than "object" and because these commands will work on two types of files, solution files and project files.

**Comment by jgoshi** [comment]
@MattGertz
